File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        freq = {}

        for num in nums:
            freq[num] = freq.get(num, 0) + 1

        buckets = [[] for i in range(len(nums) + 1)]

        for key, value in freq.items():
            buckets[value].append(key)

        topK = []
        for i in range(len(buckets) - 1, -1, -1):
            for element in buckets[i]:
                topK.append(element)

                if len(topK) == k:
                    return topK

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> numsFreq = new HashMap<>();

        for (int n: nums) {
            numsFreq.put(n, numsFreq.getOrDefault(n, 0) + 1);
        }

        Queue<Integer> heap = new PriorityQueue<>(Comparator.comparingInt(numsFreq::get));

        for (int n : numsFreq.keySet()) {
            heap.add(n);
            if (heap.size() > k) heap.poll();
        }

        int[] topK = new int[k];
        for (int i = 0; i < k; i++) {
            topK[i] = heap.poll();
        }
        return topK;
    }
}